infrastructure/000-aws-base/modules/vpc/vpc.tf

```diff
@@ -60,9 +72,57 @@ module "vpc" {
   public_subnets  = values(module.public_subnet_cidrs.network_cidr_blocks)
 
   private_subnet_ipv6_prefixes = range(local.subnet_count)
-  public_subnet_ipv6_prefixes	= range(local.subnet_count, local.subnet_count * 2)
+  public_subnet_ipv6_prefixes  = range(local.subnet_count, local.subnet_count * 2)
 
   enable_nat_gateway = true
 
   tags = merge(var.tags)
 }
+
+module "dynamodb_vpc_endpoint" {
+  source = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
+
+  vpc_id             = module.vpc.vpc_id
+  security_group_ids = [module.vpc.default_security_group_id]
+
+  endpoints = {
+    dynamodb = {
+      service         = "dynamodb"
+      service_type    = "Gateway"
+      route_table_ids = flatten([module.vpc.intra_route_table_ids, module.vpc.private_route_table_ids, module.vpc.public_route_table_ids])
+      policy          = data.aws_iam_policy_document.dynamodb_endpoint_policy.json
+    }
+  }
+
+  tags = merge(var.tags)
+}
+
+# Data source used to avoid race condition
+data "aws_vpc_endpoint_service" "dynamodb" {
+  service = "dynamodb"
+
+  filter {
+    name   = "service-type"
+    values = ["Gateway"]
+  }
+}
+
+data "aws_iam_policy_document" "dynamodb_endpoint_policy" {
+  statement {
+    effect    = "Deny"
+    actions   = ["dynamodb:*"]
+    resources = ["*"]
+
+    principals {
+      type        = "*"
+      identifiers = ["*"]
+    }
+
+    condition {
+      test     = "StringNotEquals"
+      variable = "aws:sourceVpce"
+
+      values = [data.aws_vpc_endpoint_service.dynamodb.id]
+    }
+  }
+}
```

Comment:
Quick note for those unfamiliar with the several layers of IAM policies (I was partially in this camp yesterday). This policy **only** governs who may use the VPC endpoint. As a quick reminder, policies are deny by default, approve by policy and explicit deny by policy overrides all approves. This policy has an explicit deny that says all `dynamodb` actions through this gateway will be denied if the source of those requests is from **outside** the VPC for which the endpoint is provided.

This is a very sane default, since this is the purpose of a VPC endpoint. If another VPC happens to route to this endpoint or some external IP finds its way here, it will be denied. That type of behavior should use a separate VPC endpoint or public dynamodb address.

Reply:
I briefly alluded to the "layers" of policies, but didn't actually mention all of them.

* IAM Permissions Boundary Policy - this policy is designed to give high level access or denial to resources for a user or role. Users and roles cannot get out of this boundary even if a policy below allows them. Typically these are broad and just a sanity check/failsafe for bad policies below.
* IAM Role/User Policy - These are where most people write permissions. Should a user or role be able to do something? Here's where you add it. This policy is designed for high level resource access, but where most work gets done.
* VPC Endpoint Access Policy - This policy is mainly designed around network traffic and has variables provided that relate to origin and destination.
* DynamoDB Resource Policy - This policy is data-focused. Typically this is used to restrict user/role access on rows or columns of a dynamodb table. It's possible to restrict roles wholesale, but usually makes sense for that to be done via IAM Role policies unless there are specific data-silos and ownership patterns at work.